README/README.CXXMODULES.md

```diff
@@ -0,0 +1,486 @@
+equivalent to:
+```cpp
+// ROOT prompt
+root [] namespace foo { };struct S;
+root [] foo::bar/*store parsing state*/
+        gSystem->Load("Foo");
+        // More scaffolding.
+        extern int __Cling_Autoloading_Map;
+        namespace foo{struct __attribute__((annotate("$clingAutoload$Foo.h"))) bar;}
+        struct __attribute__((annotate("$clingAutoload$Foo.h"))) S;
+        // More initialization scaffolding.
+        /*restore parsing state*/ *baz1;
+```
+
+Line #3 requires a definition and the implementation behaves exactly as in #2.
+Then it is informed that a definition is required, it reads the information in
+the annotation and parses `Foo.h`. The recursive parsing happens at two places
+making this code path error prone.
+```cpp
+// ROOT prompt
+root [] namespace foo { };struct S;
+root [] foo::bar/*store parsing state*/
+        gSystem->Load("Foo");
+        // More scaffolding.
+        extern int __Cling_Autoloading_Map;
+        namespace foo{struct __attribute__((annotate("$clingAutoload$Foo.h"))) bar;}
+        struct __attribute__((annotate("$clingAutoload$Foo.h"))) S;
+        // More initialization scaffolding.
+        /*restore parsing state*/ baz1 /*store parsing state*/
+        #include <Foo.h>/*restore parsing state*/;
+```
+
+To recap, unfortunately, ROOT PCH is not extendable; ROOTMAP require a lot of
+maintenance and goes on a very untested codepath, while RDICT has a very limited
+scope. The three features require a lot of mechanisms to work together and the
+corner cases are very many. The interaction between some of the features are
+not well defined.
+
+
+## From C++ Modules to Dictionaries
+
+C++ Modules have native capabilities to avoid reparsing. It combines all
+home-grown solutions to avoid the costly operation at industry quality.
+
+Currently, when ROOT is built with `-Druntime_cxxmodules=On` it gives priority to
+C++ Module files (real *pcm* files). If such a file is present it reads all
+necessary information from it. If no such file is present ROOT falls back to the
+standard information flow.
+
+### Observable differences from 'standard' ROOT
+
+As always, ROOT is (mostly) API and ABI compatible. C++ Modules-aware ROOT is no
+different. There are several differences which can be noticed:
+  * \*modulemap files in $ROOTSYS/include -- those files are used by rootcling to
+  put a set of header files in a single pcm file. For example, all related
+  headers of *libGeom* are persisted in *Geom.pcm*. There are a few notable
+  examples, which are specific to the way we build ROOT. In certain cases we
+  want some header files to be compiled within C context or with RTTI on/off.
+  That's mostly for bootstrapping ROOT (aka rootcling stage1).
+  * modulemap.overlay.yaml -- automatically created virtual filesystem overlay
+  file. This file contains introduces C++ Modules for third-party dependencies.
+  For example, to 'modularize' glibc for ROOT we need to place a modulemap
```

Comment:
*would* need to place. We don't actually need to.

Reply:
We do ;) but fixed.